openpype/lib/avalon_context.py

```diff
@@ -1971,3 +1971,123 @@ def get_last_workfile(
+    Args:
+        project (str)
+        repre_ids (list) or (ObjectId)
+        dbcon (avalon.mongodb.AvalonMongoDB, optional): Avalon Mongo connection
+            with Session.
+        link_type (str): ['reference', '..]
+        max_depth (int): limit how many levels of recursion
+    Returns:
+        (list) of ObjectId - linked representations
+    """
+    if not dbcon:
+        log.debug("Using `avalon.io` for query.")
+        dbcon = avalon.io
+        # Make sure is installed
+        dbcon.install()
+
+    if dbcon.Session["AVALON_PROJECT"] != project:
+        dbcon.Session["AVALON_PROJECT"] = project
+
+    if not isinstance(repre_ids, list):
+        repre_ids = [repre_ids]
+
+    versions = avalon.io.find(
+        {
+            "_id": {"$in": repre_ids},
+            "type": "representation"
+        },
+        projection={"parent": True}
+    )
+    version_ids = [version["parent"] for version in versions]
+
+    graph_lookup = {
+        "from": project,
+        "startWith": "$data.inputLinks.id",
+        "connectFromField": "data.inputLinks.id",
+        "connectToField": "_id",
+        "as": "outputs_recursive",
+        "depthField": "depth"
+    }
+    if max_depth != 0:
+        # We offset by -1 since 0 basically means no recursion
+        # but the recursion only happens after the initial lookup
+        # for outputs.
+        graph_lookup["maxDepth"] = max_depth - 1
+
+    match = {
+        "_id": {"$in": version_ids},
+        "type": "version"
+    }
+
+    pipeline_ = [
+        # Match
+        {"$match": match},
+        # Recursive graph lookup for inputs
+        {"$graphLookup": graph_lookup}
+    ]
+
+    result = dbcon.aggregate(pipeline_)
+    referenced_version_ids = _process_referenced_pipeline_result(result,
+                                                                 link_type)
+
+    representations = avalon.io.find(
+        {
+            "parent": {"$in": list(referenced_version_ids)},
+            "type": "representation"
+        },
+        projection={"_id": True}
+    )
+    ref_ids = {representation["_id"] for representation in representations}
+    return list(ref_ids)
+
+
+def _process_referenced_pipeline_result(result, link_type):
+    """Filters result from pipeline for particular link_type.
+
+    Pipeline cannot use link_type directly in a query.
+    Returns:
+        (list)
+    """
+    referenced_version_ids = set()
+    correctly_linked_ids = set()
+    for item in result:
+        correctly_linked_ids = _filter_input_links(item["data"]["inputLinks"],
+                                                   link_type,
+                                                   correctly_linked_ids)
+
+        # outputs_recursive in random order, sort by _id
+        outputs_recursive = sorted(item.get("outputs_recursive", []),
+                                   key=lambda d: d["_id"])
+        # go from oldest to newest
+        # only older _id can reference another newer _id
```

Comment:
Are we sure about this? As I refactored the Integrator I'm pretty sure existing versions and representations can be "replaced" or overwritten with new content and thus technically an earlier generated `id` might have newer content than another id.

Also, are we sure [the IDs are sortable like that as if they are timestamps?](https://stackoverflow.com/a/12100023/1838864)

Reply:
>Also, are we sure [the IDs are sortable like that as if they are timestamps?](https://stackoverflow.com/a/12100023/1838864)

I'm pretty sure that mongoIDs are timestamps, however, this will not be the case in the v4 probably, so we should keep it in mind here

Reply:
I switched that 'depth' that comes from pipeline, that should be deterministic. (Not solving v4 though.)